add information about PyCharm for writing plugin (#4313)
qgis/QGIS-Documentation: source/docs/pyqgis_developer_cookbook/plugins/ide_debugging.rst

recommendations for setting up popular IDE's for writing and debugging QGIS
Python plugins.

Useful plugins for writing Python plugins
=========================================

> Two plugins are convenient when writing Python plugins.

Reviewer: ```suggestion
Two plugins are convenient when writing Python plugins.

```

Reviewer: An empty line seems to be necessary before an rst list.

* Go to :menuselection:`Plugins --> Manage and Install plugins…`.
  You may need to enable experimental plugins.
* *Plugin reloader*. This will let you reload a plugin instead of
  having to close and restart QGIS to have the plugin reloaded.
* *First Aid*. This will let you have a remote Python console and

>   to inspect variables when a exception is raised from a plugin.

Reviewer: ```suggestion
  to inspect variables when an exception is raised from a plugin.
```

Configuring PyCharm on Ubuntu with a compiled QGIS
==================================================

PyCharm is an IDE for Python developed by JetBrains. There is a free version
called Community Edition and a paid one called Professional. You can download
PyCharm on the website: https://www.jetbrains.com/pycharm/download

We are assuming that you have compiled QGIS on Ubuntu with the given
build directory :file:`~/dev/qgis/build/master`. It's not compulsory to have

> a self compiled QGIS, but this hasn't been tested. Paths must be adapted.

Reviewer: ```suggestion
a self compiled QGIS, but only this has been tested. Paths must be adapted.
```

Reviewer: ?

> In PyCharm, in your :guilabel:`Project Properties`, :guilabel:`Project Interpreter`,

Author: Sure, I will try

> we are going to create Python Virtual environment called `QGIS`. Click

Reviewer: ```suggestion
we are going to create a Python Virtual environment called `QGIS`. Click
```

Reviewer 2: A single ` character has no particular display in html build, ie
```suggestion
we are going to create a Python Virtual environment called ``QGIS``. Click
```
And the "click" should maybe move to a step point

the small gear and then :guilabel:`Add`. Select :guilabel:`Virtualenv environment`.
Select a generic location for all your Python projects such as
:file:`~/dev/qgis/python_venv` because we will use this Python interpreter
for all our plugins. Choose a Python 3 base interpreter available on
your system and check the next two options :guilabel:`Inherit global site-packages`
and :guilabel:`Make available to all projects`.

> Click :guilabel:`OK` and then come back on the small gear, :guilabel:`Show all`.

Reviewer: ```suggestion
Click :guilabel:`OK` and you will come back on the small gear, :guilabel:`Show all`.
```

> In the new window select your new interpreter `QGIS` and click the last icon in

Reviewer: ```suggestion
In the new window, select your new interpreter `QGIS` and click the last icon in
```

> In the new window select your new interpreter `QGIS` and click the last icon in

Author: ```suggestion
In the new window select your new interpreter ``QGIS`` and click the last icon in
```

the vertical menu :guilabel:`Show paths for the selected interpreter.`

> Finally, add the following absolute path to your QGIS Python API folder

Reviewer 2: How? Where's the Python API folder? Is it an option in the gui? In which case, should be "guilabeled".

:file:`~/dev/qgis/build/master/output/python`.

> Restart PyCharm and you can start using this newly Python virtual environment

Reviewer: ```suggestion
Restart PyCharm and you can start using this new Python virtual environment
```

> for all your plugin. PyCharm will be aware of the QGIS API and also of the

Reviewer: ```suggestion
for all your plugins. PyCharm will be aware of the QGIS API and also of the
```

PyQt API if you use Qt provided by QGIS like `from qgis.PyQt.QtCore import QDir`.
The autocompletion should work and PyCharm can inspect your code.

In the professional version of PyCharm, remote debugging is working well. For the
Community edition, remote debugging is not available. You can only do debugging inside

> unit tests.

Reviewer 2: Not sure I fully understand the implications but wouldn't "remote debug or first aid be of a help here? Extending the debugging capabilities? In which case, might be worth reminding... No?

Author: Pycharm offers it's own debugger, which is different from "remote debug" (I didn't get it working, cf other message). First AID is a debugger only when an exception has been raised, so quite different too (if I'm not wrong).

Reviewer 2: OK. And PyCharm is not listed on the remote debug plugin page so my suggestion is probably not right.

Author: It's 2 different ways to have a debugger, Pycharm and the plugin called "remote debug". Pycharm has is own debugger, which can do remote debug in the professional version, and only debug in unittest or local python file in the free version.

Reviewer 2: ok

recommendations for setting up popular IDE's for writing and debugging QGIS
Python plugins.

Useful plugins for writing Python plugins
=========================================

Two plugins are convenient when writing Python plugins.

> * Go to :menuselection:`Plugins --> Manage and Install plugins…`.

Reviewer 2: Maybe remove that bullet list point. It doesn't address same issue than the next two.

Preparing QGIS
--------------

> There is some preparation to be done on QGIS itself. One plugin is of

Reviewer 2: I Don't understand why this plugin is kept here? Why not move it at the global place you were mentioning useful plugins? Afaics, this not specifically about Eclipse usage. If it's obsolete or first aid is better, then let's not mention it anymore.

Author: I was not sure about what to do here. I kept it here because it's below the `|outofdate|`. I didn't check if it's updated. I personally didn't get it working when I tried on QGIS 2, but I don't think I will use it on QGIS 3 too ... Maybe I should try to give it another try.

Reviewer 2: > ... Maybe I should try to give it another try.

Hi @Gustry. Yes please. Would be nice if you could give it a try. Thanks. I find confusing to keep it separated from the others. And if the test is still not conclusive, we'd simply remove it from the list (note that it's still experimental while the others aren't)

Author: The plugin "Remote debug" is still experimental because it is targeting only dev. Like the "Plugin reloader". I will give it a try when I have some time.

Reviewer 2: >Like the "Plugin reloader".

Not anymore (actually for a month)

Author: Oh really?after years !

Author: Oh really?after years !

interest: **Remote Debug**.

* Go to :menuselection:`Plugins --> Manage and Install plugins...`

> * Search for *Remote Debug* ( at the moment it's still experimental, so enable

Reviewer: ```suggestion
* Search for *Remote Debug* (at the moment it's still experimental, so enable
```
